flashlight/autograd/backend/cuda/Conv2D.cpp

```diff
@@ -25,25 +57,70 @@ constexpr cudnnConvolutionBwdDataAlgo_t kBwdDataDefaultAlgo =
 constexpr cudnnConvolutionBwdFilterAlgo_t kBwdFilterDefaultAlgo =
     CUDNN_CONVOLUTION_BWD_FILTER_ALGO_0;
 
+/**
+ * Sets cudnnMathType according to the mode value.
+ *
+ * @param cDesc a pointer to `ConvDescriptor` for which the math type will be
+ * set.
+ * @param kernelAudit a pointer to the KernelPerfAudit object of the kernel for
+ * which the math type is being determined.
+ */
+inline void setCudnnMathType(
+    std::shared_ptr<fl::ConvDescriptor> cDesc,
+    fl::KernelPerfAuditPtr kernelAudit) {
+  if (kernelAudit->useMixedPrecision()) {
+    auto mode = kernelAudit->mode();
+    if (mode == fl::KernelMode::F32) {
+      CUDNN_CHECK_ERR(
+          cudnnSetConvolutionMathType(cDesc->descriptor, CUDNN_DEFAULT_MATH));
+    } else if (mode == fl::KernelMode::F32_ALLOW_CONVERSION) {
+      CUDNN_CHECK_ERR(cudnnSetConvolutionMathType(
+          cDesc->descriptor, CUDNN_TENSOR_OP_MATH_ALLOW_CONVERSION));
+    } else {
+      CUDNN_CHECK_ERR(cudnnSetConvolutionMathType(
+          cDesc->descriptor, CUDNN_TENSOR_OP_MATH_ALLOW_CONVERSION));
+    }
+  } else {
+    CUDNN_CHECK_ERR(
+        cudnnSetConvolutionMathType(cDesc->descriptor, CUDNN_DEFAULT_MATH));
+  }
+}
+
 // Get the algorithm which gives best performance.
 // Since cuDNN doesn't support memory limits, we manually choose an algorithm
 // which requires less than a specific workspace size.
-template <typename T>
-T getBestAlgorithm(const std::vector<T>& algoPerfs) {
+template <typename T, typename ALGO_TYPE>
+T getBestAlgorithm(
+    const std::vector<T>& algoPerfs,
+    const u_set<ALGO_TYPE>& preferredAlgos,
+    const bool algoMustSupportF16) {
```

Comment:
As above, it's best if bool flags like this were passed as `af::dtype`s directly so that we can easily add support for other precisions and keep everything less opinionated.